### PR TITLE
Add Ingress Class to kubectl describe ingress output

### LINKED ISSUE #1160: Print ingress class name while describing ingress resource(s)
**What would you like to be added**:
Add `spec.ingressClassName` in the output of `kubectl describe ingress <ingressName/no-name-but-list>`

**Why is this needed**:
Now that ingress controllers use spec.ingressClassName to decided weather it should implement that particular ingress or not, It is important for a person maintaining the cluster to know which ingress is being implemented by which ingress controller thus class.

### PR DESCRIPTION
**Pull request by mpuckett159**
/kind feature

#### What this PR does / why we need it:
This PR adds the Ingress Class to the output of `kubectl describe ingress`

#### Which issue(s) this PR fixes:
Fixes kubernetes/kubectl#1160

#### Special notes for your reviewer:
Based on the Ingress.Spec.IngressClassName comments (linked below) it looks like this should just be blank if the name isn't specified, so I'm setting to an empty string then updating if the name is provided.
https://github.com/kubernetes/kubernetes/blob/2b091f60ca09d3d03923a26b7cd02fb3dc6c47e2/staging/src/k8s.io/api/networking/v1/types.go#L257

#### Does this PR introduce a user-facing change?
```release-note
The output of `kubectl describe ingress` now includes an IngressClass name if available
```

### CODE REVIEW

**Comment by sftim** [comment]
For the changelog, I'd write
```
The output of `kubectl describe ingress` now includes an IngressClass name if available
```

**Comment by mpuckett159** [comment]
Yup that's more accurate, thank you.

**Review by brianpursley — CHANGES_REQUESTED** [review]
Code looks good.

Can you squash your commits?
(ref: https://www.kubernetes.dev/docs/guide/github-workflow/#squash-commits)

**Comment by brianpursley** [comment]
Question: Is printing an empty string preferable to printing something like `<none>` or even omitting the field entirely from the describe output?

I don't think I have a preference, but wonder if there is already a precedent for this somewhere else in describe.

**Comment by mpuckett159** [comment]
From my reading of [the ingress docs](https://kubernetes.io/docs/concepts/services-networking/ingress/#default-ingress-class) just before the `Ingress rules` header:

> There are some ingress controllers, that work without the definition of a default IngressClass. For example, the Ingress-NGINX controller can be configured with a [flag](https://kubernetes.github.io/ingress-nginx/#what-is-the-flag-watch-ingress-without-class) --watch-ingress-without-class. It is [recommended](https://kubernetes.github.io/ingress-nginx/#i-have-only-one-instance-of-the-ingresss-nginx-controller-in-my-cluster-what-should-i-do) though, to specify the default IngressClass as shown [below](https://kubernetes.io/docs/concepts/services-networking/ingress/#default-ingress-class).

An ingress does not technically require a class, which, imo, should result in printing a blank or `<none>` line rather than being ignored in the output. Having no class is important info since it's not recommended but technically possible. Writing `<none>` by default would be a good marker and definitely has precedent in the existing describe code.

I can add that if it sounds good.

**Comment by brianpursley** [comment]
Hmm, OK.

I don’t really have much experience with ingress.  It sounds like a nil class name could indicate that it is using the default. Or in some cases nil might even be valid for that ingress type.  Maybe `<default>` would be another option, but that might not be the case for all ingress controllers.

The more I think about it, `<none>` seems wrong, because nil means use the default, not that there is no class name.

I’m sort of leaning towards how you have it currently, with an empty string.

**Comment by mpuckett159** [comment]
Nor do I unfortunately. The more I read about this the more it looks like I need to write a function to check a few things before populating a value.

[The IngressSpec API docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingressspec-v1-networking-k8s-io) show that there is also a deprecated annotation to take into account that takes precedence over the actual `IngressClassName` property value.

Seems similar to kubernetes/kubectl#1133, actually.

**Comment by mpuckett159** [comment]
I think what I've landed on after some reading and testing is the following:
1. check for setting by spec
2. check for setting by annotation and add warning if found
3. if IngressClassName is still an empty string, query the api for IngressClass objects and search for one with the default annotation and add notation it was set by cluster default

Something to note, currently when doing a `kubectl get` it returns `<none>`

```bash
❯ kc get ingress
NAME              CLASS    HOSTS              ADDRESS     PORTS   AGE
example-ingress   <none>   hello-world.info   localhost   80      27m
```

I'm not sure if it's desirable to show the inferred class name for get or just stick with what is explicitly set in the Ingress object itself in this situation, however. If we're only fetching what is explicitly set in the config then `<none>` makes sense, however it does look like it overlooks the deprecated annotation when it's used. Seems like a bit of a sharp edge, but updating `describe` to follow this logic at least seems to smooth that out slightly.

**Comment by brianpursley** [comment]
If `kubectl get ingress` just prints `<none>`, I’d be inclined to just do the same thing in describe.

I’m a little hesitant to try to reimplement the ingress class default/precedence rules inside kubectl.  It seems like this could be tricky, especially since it sounds like some ingress providers can work without a default class name.

I’m open to it, but just want to keep things simple if possible.  I understand if there is value in doing this though.

I’d like to hear from someone else who maybe has more experience than I do with ingress.  @rikatz any thoughts on this?  @eddiezane thoughts?

**Comment by rikatz** [comment]
tl;dr because I wrote too much: I think having ingress class reflecting the ingressClassName is fine :)

The part where I write too much:

IngressClass is a tricky implementation. 

Right now, we have the "deprecated" annotation, and the new ingressClassName spec field. Users still tends to use (a lot!) the annotation and even in ingress nginx we are thinking on rolling back this "deprecation". It was a really noisy change. 

From my perspective, as the field is the "official" way of saying who "owns" that ingress, it's fine to print that field content and only it. 

Now going to the "how useful this is": some controllers does not implements class, while others implement on different ways. As far as I remember a class can only be used if it exists and was created by the clusteradmin, so users cannot add random things, and it's good on a description to know "who owns that ingress". 

I have a feeling right now that this solves half the problem: we have a description, but we don't have the "feedback" of if a controller from that class properly implemented that ingress but this is another discussion on the lack of status and conditions on ingress objects and I'm done on my KEP quota this year :)

**Comment by rikatz** [comment]
Just complementing now that I read the whole thread: I wouldn't go with "how ingressA or B does". Stick with the API spec :) 

Annotation is deprecated. An ingressClassName can be empty, so either it's "none" or it's the value on the field. This is my opinion. 

Anything other than that is too specific and can lead to situations like "controller A does this way and you don't cover", etc. The API is our contract, and we should follow it.

**Comment by mpuckett159** [comment]
For easier reference re [API for Ingress](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingress-v1-networking-k8s-io):
> IngressClassName is the name of the IngressClass cluster resource. The associated IngressClass defines which controller will implement the resource. This replaces the deprecated `kubernetes.io/ingress.class` annotation. For backwards compatibility, when that annotation is set, it must be given precedence over this field. The controller may emit a warning if the field and annotation have different values. Implementations of this API should ignore Ingresses without a class specified. An IngressClass resource may be marked as default, which can be used to set a default value for this field. For more information, refer to the IngressClass documentation.

Which is how I arrived at:
> 1. check for setting by spec
> 2. check for setting by annotation and add warning if found
> 3. if IngressClassName is still an empty string, query the api for IngressClass objects and search for one with the default annotation and add notation it was set by cluster default

**Review by rikatz — CHANGES_REQUESTED** [review]
Left some reviews.

**Comment by brianpursley** [comment]
> tl;dr because I wrote too much: I think having ingress class reflecting the ingressClassName is fine :)

@rikatz Are you saying it is fine to simply display the value of ingressClassName directly, without doing any additional lookup of the default if it is nil?

I think that is where I am headed:  just show the value, if it is set, otherwise if nil then print <none>.

do you think this is good enough?

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go [hunk 1]
```diff
@@ -2731,6 +2733,70 @@ func (i *IngressDescriber) describeIngressV1beta1(ing *networkingv1beta1.Ingress
...
+	// kubernetes.io/ingress.class annotation is deprecated but takes precedence over everything, skip querying
+	// apiserver to save resources if we find the deprecated annotation as it takes precedence anyway
+	ingAnnotations := ing.GetAnnotations()
+	if val, ok := ingAnnotations["kubernetes.io/ingress.class"]; ok {
+		ingressClassName = val + " (WARNING: set by deprecated annotation)"
+	}
+	// If the ingressClassName is still blank check for cluster default Ingress Class by finding an Ingres Class
+	// with ingressclass.kubernetes.io/is-default-class annotation set to true
+	if ingressClassName == "" {
+		icl, err := i.client.NetworkingV1().IngressClasses().List(context.TODO(), metav1.ListOptions{})
```
[the remark this answers is not in the document]
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go:2784]: Agreed.  If there is an error, you need to probably show something here, or else it could be misleading.

**Comment by rikatz** [comment]
> @rikatz Are you saying it is fine to simply display the value of ingressClassName directly, without doing any additional lookup of the default if it is nil?
> 

Yeap, I'm a +1 to keep it simple right now. It's still confusing for users this annotations vs ingressClass thing, so let's state that ingressClass is the ingressClassName field, and IMHO move on :)

**Comment by mpuckett159** [comment]
I just wanted to double check, because I feel like I'm not on the same page as everyone else, but from my understanding of the API docs, it sounds to me like the annotation takes precedence, so if we really wanted to reflect reality back to the user it seems like we should be showing that annotation back if it exists as that's what the ultimate applied class will be.

**Comment by mpuckett159** [comment]
After sleeping on it I think what everyone is saying is that k8s.io/api/networking/v1(beta1) is where we should really implement the logic that I've written and have a function made available so that anyone can figure out the actual applied IngressClassName. That sounds correct to me, and if that's the consensus, I can scale back this commit back to where it's just reading the IngressClassName property and make an issue/PR for adding a function to "calculate" the actual applied Ingress Class based on the outlined logic from the docs, and whenever that's pulled in we can update the describe code here to utilize that function properly.

**Comment by mpuckett159** [comment]
Alright all that makes sense to me now, thanks everyone for the input. I've basically reset to the original PR commits and updated the Ingress Class output to show `<none>` to ensure that it is explicitly stated that when `ing.Spec.IngressClassName` is a nil pointer we are explicitly showing the user that, and hopefully avoiding any potential confusion there. I'll work on seeing if my proposed function for calculating the actual applied Ingress Class Name for the API is acceptable or not and get something open to which ever repo I need to open it to if so.

**Review by brianpursley — APPROVED** [review]
/lgtm

**Comment by brianpursley** [comment]
Thanks @mpuckett159 👍
